### DISCUSSION TITLE
Any way to stop streaming subscription early?

### DISCUSSION THREAD

**Discussion by vectar7** [ecederstrand/exchangelib#1108] [Q&A] [answered]
I'm using exchangelib to create a Windows service that monitors a mailbox using a streaming subscription. Everything works great, except that because the streaming subscription blocks for the duration of the connection_timeout, the service will not stop when requested until the connection_timeout expires.

Is there any way that I can force the connection to end early, maybe based on a flag I set or something like that? I'm doing something like this:

```
for notification in a.inbox.get_streaming_events(subscription_id, connection_timeout=5):
    for event in notification.events:
        if isinstance(event, (NewMailEvent)):
            #do stuff
```

**Comment by ecederstrand** [comment, marked as answer]
I don't know how this sort of thing works in Windows. On Unix you would send the service process a signal, SIGHUP for example, and in the code for the service you register a signal handler that exits the process.

You could also run your for loop in a thread and have the outer process detect when to exit.

Anyway, this is a general problem and not something specific to exchangelib.

**Reply by vectar7** [reply]
I'll do some digging and see what I can come up with. I appreciate the quick response.

**Comment by vectar7** [comment]
Just in case anyone else stumbles across this with the same question...

I just ran my loop in a thread and that worked great to solve this problem. Thanks for the suggestion!

**Reply by kostrykin** [reply]
So you just killed the thread, when you needed to terminate the connection/subscription?

**Reply by ecederstrand** [reply]
Yes. You kill the thread and then unsubscribe from the subscription before exiting the process.
